source/fuzz/transformation_merge_function_returns.cpp

```diff
@@ -0,0 +1,646 @@
+      if (!info.is_returning_id() ||
+          !CheckIdIsFreshAndNotUsedByThisTransformation(
+              info.is_returning_id(), ir_context, &used_fresh_ids)) {
+        return false;
+      }
+
+      if (!function_type->AsVoid() &&
+          (!info.maybe_return_val_id() ||
+           !CheckIdIsFreshAndNotUsedByThisTransformation(
+               info.maybe_return_val_id(), ir_context, &used_fresh_ids))) {
+        return false;
+      }
+
+      // Get the mapping from OpPhis to suitable ids.
+      phi_to_id = fuzzerutil::RepeatedUInt32PairToMap(
+          *info.mutable_opphi_to_suitable_id());
+    } else {
+      // If the map does not contain an entry for the merge block, check that
+      // overflow ids are available.
+      if (!transformation_context.GetOverflowIdSource()->HasOverflowIds()) {
+        return false;
+      }
+    }
+
+    // For each OpPhi instruction, check that a suitable placeholder id is
+    // available.
+    bool suitable_info_for_phi =
+        ir_context->get_instr_block(merge_block)
+            ->WhileEachPhiInst([ir_context, &phi_to_id,
+                                &types_to_available_ids](
+                                   opt::Instruction* inst) {
+              if (phi_to_id.count(inst->result_id()) > 0) {
+                // If there exists a mapping for this instruction and the
+                // placeholder id exists in the module, check that it has the
+                // correct type and it is available before the instruction.
+                auto placeholder_def = ir_context->get_def_use_mgr()->GetDef(
+                    phi_to_id[inst->result_id()]);
+                if (placeholder_def) {
+                  if (inst->type_id() != placeholder_def->type_id()) {
+                    return false;
+                  }
+                  if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+                          ir_context, inst, placeholder_def->result_id())) {
+                    return false;
+                  }
+
+                  return true;
+                }
+              }
+
+              // If there is no mapping, check if there is a suitable id
+              // available at the end of the entry block.
+              return types_to_available_ids.count(inst->type_id()) > 0;
+            });
+
+    if (!suitable_info_for_phi) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void TransformationMergeFunctionReturns::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  auto function = ir_context->GetFunction(message_.function_id());
+  auto function_type = ir_context->get_type_mgr()->GetType(function->type_id());
+
+  // Get a map from the types for which ids are available at the end of the
+  // entry block to one of the ids with that type. We compute this here to avoid
+  // potentially doing it multiple times later on.
+  auto types_to_available_ids =
+      GetTypesToIdAvailableAfterEntryBlock(ir_context);
```

Comment:
Good - I like the way this is at the start; I'd do something similar in IsApplicable above.

Reply:
In the IsApplicable function, I have moved it almost to the start, just after the check that the function actually exists